windows/src/desktop/kmshell/main/OnlineUpdateCheck.pas

```diff
@@ -706,4 +975,426 @@ function TOnlineUpdateSharedData.Params: TOnlineUpdateCheckParams;
   Result := FParams;
 end;
 
+
+function SetBackgroundState(Update : TUpdateState): Boolean;  // I2329
+var
+  UpdateStr : string;
+begin
+
+  Result := False;
+  // TODO Ask Marc how to handle the execption and return false
+  with TRegistryErrorControlled.Create do  // I2890
+  try
+    RootKey := HKEY_LOCAL_MACHINE;
+    if OpenKey(SRegKey_KeymanEngine_LM, True) then
+    begin
+        UpdateStr := GetEnumName(TypeInfo(TUpdateState), Ord(Update));
+        Write(SRegValue_Update_State, UpdateStr);
+    end;
+    Result := True;
+  //except
+  //  on E:Exception do
+  //  begin
+   //   Result := False;
+  //  end;
+  finally
+      Free;
+  end;
+
+end;
+
+function CheckBackgroundState : TUpdateState;  // I2329
+var
+  UpdateState : TUpdateState;
+
+begin
+  // We will use a registry flag to maintain the state of the background update
+
+  UpdateState := Idle; // do we need a unknown state ?
+  // check the registry value
+  with TRegistryErrorControlled.Create do  // I2890
+  try
+    RootKey := HKEY_LOCAL_MACHINE;
+    if OpenKeyReadOnly(SRegKey_KeymanEngine_LM) and ValueExists(SRegValue_Update_State) then
+        try
+          UpdateState := TUpdateState(GetEnumValue(TypeInfo(TUpdateState), ReadString(SRegValue_Update_State)));
+          KL.Log('CheckBackgroundState State is:[' + ReadString(SRegValue_Update_State) + ']');
+        except
+          UpdateState := Idle; // do we need a unknown state ?
+        end;
+    finally
+      Free;
+  end;
+  Result := UpdateState;
+end;
+
+function TOnlineUpdateCheck.BackgroundInstall : Boolean;  // I2329
+var
+  Update : Boolean;
+  SavePath: String;
+  fileExt : String;
+  fileName: String;
+  fileNames: TStringDynArray;
+begin
+    // If Keyman is failed to be shutdown then it will exit
+    // the state will remain in background for the next time processbackgoundinstall is called.
+    SavePath := IncludeTrailingPathDelimiter(GetFolderPath(CSIDL_COMMON_APPDATA) + SFolder_CachedUpdateFiles);
+    GetFileNamesInDirectory(SavePath, fileNames);
+    // for now we only want the exe all though excute install can
+    // handle msi and msp
+    for fileName in fileNames do
+    begin
+      fileExt := LowerCase(ExtractFileExt(fileName));
+      if fileExt = '.exe' then
+        break;
+    end;
+    // ExecuteInstall(SavePath + ExtractFileName(fileName));
+    // TODO DoInstallPackages ( this may need to be as state in the enum seperate
+    // to installing the main keyman executable.
+    Result := DoInstallKeyman(SavePath + ExtractFileName(fileName));
+end;
+
+function TOnlineUpdateCheck.IsKeymanRunning: Boolean;  // I2329
+begin
+  try
+      Result :=  kmcom.Control.IsKeymanRunning;
+  except
+    on E:Exception do
+    begin
+      KL.Log(E.Message);
+      Exit;
```

Comment:
Need an exit value here, assume false:

```suggestion
      Exit(False);
```

But I'm also not a big fan of catch-all exceptions. We should only be catching expected exceptions really -- is this wrapper for a reason?